### PR TITLE
retry: keep retrying on AppGW until access is granted 

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

This PR makes AGIC wait forever until it gets access to Application Gateway.
1. added a WaitForGetAccessOnGateway
1. updated configmap to take clientId provided in the aad pod identity. This information will be used to provide full az cli command to use if permission is missing.
1. changed retry function to take -1 as forever.
1. added e2e using role assignment client.

Closed: https://github.com/Azure/application-gateway-kubernetes-ingress/pull/755
I have closed this PR after realising that role assignments are created using objectIDs and AGIC only has clientId which cannot be used to determine if AGIC has contributor access to App Gateway or not. When we try to list the role assignments, we get the whole list of assignments and we cannot filter this list without objectId. Asking users to provide objectId would help but that would increase the variables we are asking the user to provide and making this mandatory might make it harder to upgrade.

### CODE REVIEW

**Comment by akshaysngupta** [comment]
E2E passing:
https://dev.azure.com/azure/application-gateway-kubernetes-ingress/_releaseProgress?releaseId=290&_a=release-environment-extension&environmentId=576&extensionId=ms.vss-test-web.test-result-in-release-environment-editor-tab

File: pkg/azure/client.go [hunk 1]
```diff
@@ -118,50 +114,76 @@ func (az *azClient) SetAuthorizer(authorizer autorest.Authorizer) {
...
+				if response.Response.StatusCode == 403 {
+					e.Code = controllererrors.ErrorApplicationGatewayForbidden
+
+					clientID := "<agic-client-id>"
+					if az.clientID != "" {
+						clientID = az.clientID
+					}
+
+					e.Message += fmt.Sprintf(
+						" You can use 'az role assignment create --role Contributor --scope /subscriptions/%s/resourceGroups/%s --assignee %s' to assign permissions.",
```
Comment (3quanfeng) [inline pkg/azure/client.go:152]: Should the scope be App-Gateway-ID instead of resource group? resource group only needs reader role

File: pkg/environment/environment.go [hunk 2]
```diff
@@ -20,6 +20,9 @@ const (
 	// CloudProviderConfigLocationVarName is an environment variable name. This file is available on azure cluster.
 	CloudProviderConfigLocationVarName = "AZURE_CLOUD_PROVIDER_LOCATION"
 
+	// ClientIDVarName is an environment variable which stores the client id provided through user assigned identity
+	ClientIDVarName = "AZURE_CLIENT_ID"
```
Comment (3quanfeng) [inline pkg/environment/environment.go:24]: AZURE_UMI_CLIENT_ID?

File: cmd/appgw-ingress/main.go [hunk 3]
```diff
@@ -138,24 +138,27 @@ func main() {
 		azClient.SetAuthorizer(authorizer)
 	}
 
-	if _, err = azClient.GetGateway(); err != nil {
-		if controllererrors.IsErrorCode(err, controllererrors.ErrorApplicationGatewayNotFound) &&
-			env.EnableDeployAppGateway {
+	// check if application gateway exists/have get access
+	// When an AAD SP doesn't have access to the app gateway resource group, then AGIC can't read it's role assignments to look for the needed permission.
```
Comment (3quanfeng) [inline cmd/appgw-ingress/main.go:142]: nit: When an AAD SP doesn't have read access to the app gateway resource group...

File: pkg/environment/environment.go (hunk 2 above)
Comment (akshaysngupta) [reply pkg/environment/environment.go:24]: This is env variable defined the by Azure Go SDK. So we have to use this string.
https://github.com/Azure/azure-sdk-for-go#more-authentication-details

File: pkg/environment/environment.go [hunk 4]
```diff
@@ -109,7 +113,7 @@ type EnvVariables struct {
 	AppGwResourceID             string
 	AppGwSubnetID               string
 	AuthLocation                string
-	IngressClass               string
+	IngressClass                string
```
Comment (3quanfeng) [inline pkg/environment/environment.go]: unrelated change?

File: pkg/azure/client.go (hunk 1 above)
Comment (akshaysngupta) [reply pkg/azure/client.go:152]: 1) Either we can add this:
```
az role assignment create --role Reader --scope /subscriptions/%s/resourceGroups/%s --assignee %s
az role assignment create --role Contributor --scope <appGatewayId> --assignee %s
```

2) Or we can aim for Contributor to the resource group
a) One command to give access
b) Might be helpful when vnet is in the same resource group. that way, we can use this access to apply route table to the subnet through AGIC without asking for addtitional permission to the Virtual network.

Comment (3quanfeng) [reply pkg/azure/client.go:152]: I think we should go with 1st option, the role assignment should be granted as least as possible unless the contributor role becomes mandatory

File: scripts/e2e/cmd/runner/environment.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		SubscriptionID:     os.Getenv(SubscriptionIDVarName),
+		ResourceGroupName:  os.Getenv(ResourceGroupNameVarName),
+		AppGwName:          os.Getenv(AppGwNameVarName),
+		KubeConfigFilePath: GetEnvironmentVariable(KubeConfigVarName, "~/.kube/config", nil),
+		ObjectID:           os.Getenv(ObjectIDVarName),
+	}
+}
+
+// GetGroupResourceID returns group's resource id
+func (env *EnvVariables) GetGroupResourceID() string {
```
Comment (3quanfeng) [inline scripts/e2e/cmd/runner/environment.go:54]: GetResourceGroupID()?

File: scripts/e2e/cmd/runner/mfu_one_namespace_one_ingress_test.go [hunk 6]
```diff
@@ -129,6 +129,66 @@ var _ = Describe("MFU", func() {
...
+
+			// delete the AGIC pod. This will create the pod
+			klog.Info("Deleting AGIC pod")
+			err = deleteAGICPod(clientset)
+			Expect(err).To(BeNil())
+
+			// wait for 30 seconds
+			klog.Info("Wait for 60 seconds")
+			time.Sleep(60 * time.Second)
+
```
Comment (3quanfeng) [inline scripts/e2e/cmd/runner/mfu_one_namespace_one_ingress_test.go:159]: how can we check if AGIC pod is in ready status?

Comment (akshaysngupta) [reply scripts/e2e/cmd/runner/mfu_one_namespace_one_ingress_test.go:159]: let me address this in a separate PR

Comment (3quanfeng) [reply scripts/e2e/cmd/runner/mfu_one_namespace_one_ingress_test.go:159]: Sounds good, please help track with a task
